sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3313,20 +3313,34 @@ class Dataset[T] private[sql](
             }
           }
 
-        val arrowBatchRdd = toArrowBatchRdd(plan)
-        sparkSession.sparkContext.runJob(
-          arrowBatchRdd,
-          (it: Iterator[Array[Byte]]) => it.toArray,
-          handlePartitionBatches)
+        var sparkException: Option[SparkException] = None
+        try {
+          val arrowBatchRdd = toArrowBatchRdd(plan)
+          sparkSession.sparkContext.runJob(
+            arrowBatchRdd,
+            (it: Iterator[Array[Byte]]) => it.toArray,
+            handlePartitionBatches)
+        } catch {
+          case e: SparkException =>
+            sparkException = Some(e)
+        }
 
-        // After processing all partitions, end the stream and write batch order indices
+        // After processing all partitions, end the batch stream
         batchWriter.end()
```

Comment:
I'm wondering if this and the code below should be in a finally block?

Reply:
If we put it into a finally block but only catch `SparkException` then that would be wrong: If a different exception gets thrown then we would go into `case None`, end the stream as if nothing happened and only get partial, incorrect data on the python side.
If we want to put this into a finally block then we should catch all exceptions but I figured I'd do the same as in https://github.com/apache/spark/pull/24070/files#r279589039

It should be fine as is, if any exception that isn't a `SparkException` gets thrown then we will never reach this code. Instead the `OutputStream` just gets closed and we get an `EofError` on the python side (like we do right now for all Exceptions).

Reply:
Any more thoughts on this @BryanCutler ?

Reply:
Yeah, I think this is fine